### PR TITLE
Firefox try support

### PR DESCRIPTION
**Pull request by EricRahm**
Adds support for downloading Firefox try builds. The implementation and tests are primarily based on TinderboxScraper. 

Example usage: |mozdownload -t try --changeset=635d76cdfe1e|

This fixes issue #239

### CODE REVIEW

**Comment by EricRahm** [comment]
@Nebelhom would you mind looking at this? @whimboo is pretty busy for the next week or so.

File: mozdownload/parser.py [hunk 1]
```diff
@@ -56,6 +56,7 @@ def handle_data(self, data):
         if not self.active_url:
             return
 
-        name = urllib.quote(data.strip('/'))
```
Comment (Nebelhom) [inline mozdownload/parser.py]: Why was this change necessary?

File: mozdownload/scraper.py [hunk 2]
```diff
@@ -870,13 +870,130 @@ def platform_regex(self):
...
+        self.changeset = changeset
+
+        Scraper.__init__(self, *args, **kwargs)
+
+    def get_build_info(self):
+        "Defines additional build information"
+
+        self.build_index = None
+
+        self.locale_build = self.locale != 'en-US'
```
Comment (Nebelhom) [inline mozdownload/scraper.py]: I don't like expressions like this, as I feel python should be about readability (This is a personal preference).

On a separate note, except for b2g builds (where it is `multi`), the default locale is en-US anways. What is the reason for the last part of this line?

File: mozdownload/scraper.py [hunk 3]
```diff
@@ -870,13 +870,130 @@ def platform_regex(self):
...
+
+    def get_build_info_for_index(self, build_index=None):
+        url = urljoin(self.base_url, self.build_list_regex)
+
+        self.logger.info('Retrieving list of builds from %s' % url)
+        parser = DirectoryParser(url, authentication=self.authentication,
+                                 timeout=self.timeout_network)
+        parser.entries = parser.filter('.*-%s$' % self.changeset)
+
+        if not parser.entries:
```
Comment (Nebelhom) [inline mozdownload/scraper.py]: Do you ever just download a `latest` build and there is a chance that the folder is empty and a `NotFoundError` may happen?

If that is the case you may want to have a look at issue #11 and its solution. I am currently implementing this for tinderbox, too, where this may happen.

File: tests/try_scraper/test_invalid_changeset.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+from mozdownload.utils import urljoin
+
+import mozhttpd_base_test as mhttpd
+
+
+tests_with_invalid_changeset = [
+    # -a firefox -p linux --changeset 97322d2d2d6c
+    {'args': {'application': 'firefox',
+              'changeset': '97322d2d2d6c',
+              'platform': 'win32'}
```
Comment (Nebelhom) [inline tests/try_scraper/test_invalid_changeset.py]: Discrepancy between comment and test parameter. Change one or the other.

File: tests/try_scraper/test_invalid_changeset.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    },
```
Comment (Nebelhom) [inline tests/try_scraper/test_invalid_changeset.py]: Any particular reason why you do not compare `target_url` and `target` to check that you are downloading the correct file?

File: tests/try_scraper/test_invalid_changeset.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+]
+
+tests = tests_with_invalid_changeset
+
+
+class TestTinderboxScraper_invalidParameters(mhttpd.MozHttpdBaseTest):
```
Comment (Nebelhom) [inline tests/try_scraper/test_invalid_changeset.py]: This should be called `TestTryScraper`

File: tests/try_scraper/test_invalid_changeset.py (hunk 5 above)
Comment (Nebelhom) [inline tests/try_scraper/test_invalid_changeset.py]: Ignore the comment above ;)

File: tests/try_scraper/test_try_scraper.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+import os
+import unittest
+import urllib
+
+from mozdownload import TryScraper
+from mozdownload.utils import urljoin
+import mozhttpd_base_test as mhttpd
+
+firefox_tests = [
```
Comment (Nebelhom) [inline tests/try_scraper/test_try_scraper.py]: Could you please put the parameters for `args` onto a separate line each?

For readability reasons and to be inline with the other tests.

**Comment by Nebelhom** [comment]
@EricRahm 

I just double-checked your PR and I did not find any major issue.

I am not familiar with the try builds , but code-wise apart from the comments and minor changes I suggested, you would get a +1.

For merging the changes, however, we will have to wait for @whimboo as I do not have admin rights.

One more thing to note. Once the PR is accepted, prior to merging the commits need to get squashed into one and the commit message should be a summary of the changes with the issue it fixes in brackets at the end like so `(#123)`

Thanks for the addition! :) Good stuff! Keep up the good work :)

File: mozdownload/parser.py (hunk 1 above)
Comment (EricRahm) [inline mozdownload/parser.py]: Try entries have the form `user@domain-<changeset>`. It appears some servers escape the `@` (mozhttpd which our unit tests use does) and some don't (http://ftp.mozilla.org/pub/mozilla.org/firefox/try-builds/).

File: mozdownload/scraper.py (hunk 2 above)
Comment (EricRahm) [inline mozdownload/scraper.py]: I followed the pattern that [TinderboxScraper uses](https://github.com/mozilla/mozdownload/blob/master/mozdownload/scraper.py#L744), in retrospect I don't think it is necessary for try builds, I'll remove it.

File: mozdownload/scraper.py (hunk 3 above)
Comment (EricRahm) [inline mozdownload/scraper.py]: Try builds don't have a concept of `latest`, so we luckily don't have to worry about that here.

File: tests/try_scraper/test_invalid_changeset.py (hunk 4 above)
Comment (EricRahm) [inline tests/try_scraper/test_invalid_changeset.py]: Will update.

File: tests/try_scraper/test_invalid_changeset.py (hunk 6 above)
Comment (EricRahm) [inline tests/try_scraper/test_invalid_changeset.py]: Will update.

File: tests/try_scraper/test_try_scraper.py (hunk 7 above)
Comment (EricRahm) [inline tests/try_scraper/test_try_scraper.py]: Sounds reasonable, I'll clean it up.

**Comment by EricRahm** [comment]
@Nebelhom thanks for the review! I'll get a cleaned up push together. Do you prefer I push the squashed commit to a new branch or just just `push -f` to this branch?

**Comment by EricRahm** [comment]
Updated and squashed.

File: mozdownload/parser.py (hunk 1 above)
Comment (Nebelhom) [inline mozdownload/parser.py]: Interesting. I'll try to remember that. Thanks for the clarification :)

**Comment by Nebelhom** [comment]
@EricRahm Great stuff! +1 from me.

Let's wait for @whimboo . I hope I found most of the nitpicks, so that it should be easier for him :)

Thanks for the PR!

File: mozdownload/parser.py [hunk 8]
```diff
@@ -56,6 +56,7 @@ def handle_data(self, data):
...
-        if self.active_url == name:
+        name = data.strip('/')
```
Comment (whimboo) [inline mozdownload/parser.py]: Lets get rid of stripping the final slash from here and `handle_starttag`. Instead lets do it when we call `self.entries.append`. That way we do not need the extra two variables (name, name_quoted) defined here. Lets pass this directly into the tuple.

File: mozdownload/scraper.py [hunk 9]
```diff
@@ -870,13 +870,126 @@ def platform_regex(self):
         return PLATFORM_FRAGMENTS[self.platform]
 
 
+class TryScraper(Scraper):
+    "Class to download a try build from the Mozilla server."
+
+    def __init__(self, changeset=None,
+                 debug_build=False, *args, **kwargs):
+
+        self.debug_build = debug_build
```
Comment (whimboo) [inline mozdownload/scraper.py]: nit: put all parameters into a single line, and kill the above empty one.

File: mozdownload/scraper.py [hunk 10]
```diff
@@ -870,13 +870,126 @@ def platform_regex(self):
...
+                       self.builds[self.build_index],
+                       try_dir)
+
+    @property
+    def platform_regex(self):
+        """Return the platform fragment of the URL"""
+
+        PLATFORM_FRAGMENTS = {'linux': 'linux',
+                              'linux64': 'linux64',
+                              'mac': 'macosx',
```
Comment (whimboo) [inline mozdownload/scraper.py]: I think you want to let mac point to ´macosx64´ too.

File: mozdownload/scraper.py [hunk 11]
```diff
@@ -870,13 +870,126 @@ def platform_regex(self):
...
+                              'mac64': 'macosx64',
+                              'win32': 'win32',
+                              'win64': 'win64'}
```
Comment (whimboo) [inline mozdownload/scraper.py]: I have a problem here with downloading such a build for win64. Please check yourself with changeset 3e461792bb3c.

File: mozdownload/scraper.py [hunk 12]
```diff
@@ -870,13 +870,126 @@ def platform_regex(self):
...
+        url = urljoin(self.base_url, self.build_list_regex)
+
+        self.logger.info('Retrieving list of builds from %s' % url)
+        parser = DirectoryParser(url, authentication=self.authentication,
+                                 timeout=self.timeout_network)
+        parser.entries = parser.filter('.*-%s$' % self.changeset)
+
+        if not parser.entries:
+            message = 'No builds have been found'
+            raise NotFoundError(message, url)
```
Comment (whimboo) [inline mozdownload/scraper.py]: Please add the message directly to the NotFoundError call. We may want to refactor some old code too later.

File: mozdownload/scraper.py [hunk 13]
```diff
@@ -870,13 +870,126 @@ def platform_regex(self):
...
+
+        self.show_matching_builds(parser.entries)
+
+        # If no index has been given, set it to the last build of the day.
+        if build_index is None:
+            build_index = len(parser.entries) - 1
```
Comment (whimboo) [inline mozdownload/scraper.py]: So we do not have any index information for try server builds because each changeset corresponds to exactly one sub folder. So build_index will always be None here, right? I would suggest we hard-code 0 in line 956 and 958.

File: mozdownload/scraper.py [hunk 14]
```diff
@@ -870,13 +870,126 @@ def platform_regex(self):
...
+        self.logger.info('Selected build: %s' % parser.entries[build_index])
+
+        return (parser.entries, build_index)
+
+    @property
+    def path_regex(self):
+        """Return the regex for the path"""
+
+        try_dir = 'try-%(PLATFORM)s%(DEBUG)s' % {
```
Comment (whimboo) [inline mozdownload/scraper.py]: I would call the variable `build_dir`.

File: mozdownload/scraper.py [hunk 15]
```diff
@@ -870,13 +870,126 @@ def platform_regex(self):
...
+
 def cli():
     """Main function for the downloader"""
 
     BUILD_TYPES = {'release': ReleaseScraper,
                    'candidate': ReleaseCandidateScraper,
                    'daily': DailyScraper,
-                   'tinderbox': TinderboxScraper}
+                   'tinderbox': TinderboxScraper,
+                   'try': TryScraper}
```
Comment (whimboo) [inline mozdownload/scraper.py]: While we are at it please also add a comma after TryScraper and move the closing bracket to the next line.

File: mozdownload/scraper.py [hunk 16]
```diff
@@ -1008,6 +1121,14 @@ def cli():
                      help="Download a debug build")
     parser.add_option_group(group)
 
+    # Option group for try builds
+    group = OptionGroup(parser, "Try builds",
+                        "Extra options for try builds.")
+    group.add_option('--changeset',
+                     dest='changeset',
+                     help="Changeset of the try build to download")
```
Comment (whimboo) [inline mozdownload/scraper.py]: Lets try to stick with single quote signs in all the cases here. We might have to do a clean-up for older code in the future.

File: mozdownload/scraper.py [hunk 17]
```diff
@@ -1051,7 +1172,9 @@ def cli():
         'tinderbox': {'branch': options.branch,
                       'build_number': options.build_number,
                       'date': options.date,
-                      'debug_build': options.debug_build}
+                      'debug_build': options.debug_build},
+        'try': {'changeset': options.changeset,
+                'debug_build': options.debug_build}
```
Comment (whimboo) [inline mozdownload/scraper.py]: Add a trailing comma here too.

File: tests/try_scraper/test_invalid_changeset.py [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import os
+import mozlog
+import sys
+import unittest
+import urllib
```
Comment (whimboo) [inline tests/try_scraper/test_invalid_changeset.py]: Lots of those imports are not used. Please get rid of them.

File: tests/try_scraper/test_invalid_changeset.py [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+from mozdownload import NotFoundError
+from mozdownload.utils import urljoin
+
+import mozhttpd_base_test as mhttpd
+
+
+tests_with_invalid_changeset = [
+    # -a firefox -p win32 --changeset 97322d2d2d6c
+    {'args': {'application': 'firefox',
+              'changeset': '97322d2d2d6c',
```
Comment (whimboo) [inline tests/try_scraper/test_invalid_changeset.py]: The risk is low but it could still be a valid changeset. Lets specify only 4 characters like 'abcd'.

File: tests/try_scraper/test_try_scraper.py [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+     'target': '8fcac92cfcad-firefox-38.0a1.en-US.linux-x86_64.tar.bz2',
+     'target_url': 'firefox/try-builds/test-user@mozilla.com-8fcac92cfcad/'
+                   'try-linux64/firefox-38.0a1.en-US.linux-x86_64.tar.bz2'},
+    # -a firefox -p linux --debug-build
+    {'args': {'platform': 'linux',
+              'changeset': '8fcac92cfcad',
+              'debug_build': True},
+     'target': '8fcac92cfcad-debug-firefox-38.0a1.en-US.linux-i686.tar.bz2',
+     'target_url': 'firefox/try-builds/test-user@mozilla.com-8fcac92cfcad/'
+                   'try-linux-debug/firefox-38.0a1.en-US.linux-i686.tar.bz2'},
```
Comment (whimboo) [inline tests/try_scraper/test_try_scraper.py]: I miss tests for win32 and win64.

File: tests/try_scraper/test_try_scraper.py [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+import os
+import unittest
+import urllib
+
+from mozdownload import TryScraper
+from mozdownload.utils import urljoin
+import mozhttpd_base_test as mhttpd
+
+firefox_tests = [
+    # -p mac64 --changeset=8fcac92cfcad
```
Comment (whimboo) [inline tests/try_scraper/test_try_scraper.py]: You also want to test `mac` as platform.

File: tests/try_scraper/test_try_scraper.py [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+class TryScraperTest(mhttpd.MozHttpdBaseTest):
+    """test mozdownload TryScraper class"""
+
+    def test_scraper(self):
+        """Testing various download scenarios for TryScraper"""
+
+        for entry in tests:
+            scraper = TryScraper(directory=self.temp_dir, version=None,
+                                       base_url=self.wdir, log_level='ERROR',
+                                       **entry['args'])
```
Comment (whimboo) [inline tests/try_scraper/test_try_scraper.py]: There is a broken indentation in those two lines.

**Comment by whimboo** [comment]
Fantastic work Eric! This looks great. Please find all my comments inline. Let me know if something is unclear. If you can please fix all those ASAP given that we have to release a new version of mozdownload soon, and I really would like to see it included.

File: mozdownload/scraper.py (hunk 10 above)
Comment (EricRahm) [inline mozdownload/scraper.py]: |detect_platform| does some special casing for this, I believe that should be enough. Although maybe that's changed?

File: mozdownload/scraper.py (hunk 11 above)
Comment (EricRahm) [inline mozdownload/scraper.py]: Fun so win64 builds can have two forms on try (see ca82d1eb3194):
- win64
- win64-x86_64

File: mozdownload/scraper.py (hunk 10 above)
Comment (EricRahm) [inline mozdownload/scraper.py]: Sorry, checked again and you're correct.

**Comment by EricRahm** [comment]
Updated, squashed, pep8'd.

File: mozdownload/parser.py (hunk 8 above)
Comment (EricRahm) [inline mozdownload/parser.py]: Done.

File: mozdownload/scraper.py (hunk 9 above)
Comment (EricRahm) [inline mozdownload/scraper.py]: Done.

File: mozdownload/scraper.py (hunk 12 above)
Comment (EricRahm) [inline mozdownload/scraper.py]: Done.

File: mozdownload/scraper.py (hunk 13 above)
Comment (EricRahm) [inline mozdownload/scraper.py]: Done. Cleaned up the call to get_build_for_index as well.

File: mozdownload/scraper.py (hunk 14 above)
Comment (EricRahm) [inline mozdownload/scraper.py]: Done.

File: mozdownload/scraper.py (hunk 15 above)
Comment (EricRahm) [inline mozdownload/scraper.py]: Done.

File: mozdownload/scraper.py (hunk 16 above)
Comment (EricRahm) [inline mozdownload/scraper.py]: Done.

File: mozdownload/scraper.py (hunk 17 above)
Comment (EricRahm) [inline mozdownload/scraper.py]: Done.

File: tests/try_scraper/test_invalid_changeset.py (hunk 18 above)
Comment (EricRahm) [inline tests/try_scraper/test_invalid_changeset.py]: Done.

File: tests/try_scraper/test_invalid_changeset.py (hunk 19 above)
Comment (EricRahm) [inline tests/try_scraper/test_invalid_changeset.py]: Done.

File: tests/try_scraper/test_try_scraper.py (hunk 21 above)
Comment (EricRahm) [inline tests/try_scraper/test_try_scraper.py]: Done. Updated test file as well.

File: tests/try_scraper/test_try_scraper.py (hunk 20 above)
Comment (EricRahm) [inline tests/try_scraper/test_try_scraper.py]: Added.

File: tests/try_scraper/test_try_scraper.py (hunk 22 above)
Comment (EricRahm) [inline tests/try_scraper/test_try_scraper.py]: Done.

File: mozdownload/scraper.py (hunk 10 above)
Comment (whimboo) [inline mozdownload/scraper.py]: > |detect_platform| does some special casing for this, I believe that
> should be enough. Although maybe that's changed?

That is only called in case of no platform has been specified via the -p
option. So if you explicitly pass in -p mac it would not work.

**Comment by whimboo** [comment]
@EricRahm when you update a PR please never squash your commits! It makes it very hard to see those lines which have been changed. I have to go through the whole file again now. :/ Thanks.

File: mozdownload/scraper.py [hunk 23]
```diff
@@ -46,7 +46,7 @@
                       'mac': r'mac',
                       'mac64': r'mac(64)?',
                       'win32': r'win32',
-                      'win64': r'win64-x86_64'}
+                      'win64': r'win64(-x86_64)?'}
```
Comment (whimboo) [inline mozdownload/scraper.py]: This was actually part of a PR I have landed yesterday. So this should cause a merge conflict.

File: mozdownload/scraper.py (hunk 11 above)
Comment (whimboo) [inline mozdownload/scraper.py]: Yes, they made this change a couple of days ago. I landed a PR yesterday with support for this new naming schema. So your latest update works perfectly.

File: tests/try_scraper/test_invalid_changeset.py [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from mozdownload import TryScraper
+from mozdownload import NotFoundError
+
+import mozhttpd_base_test as mhttpd
+
+
+tests_with_invalid_changeset = [
+    # -a firefox -p win32 --changeset 97322d2d2d6c
```
Comment (whimboo) [inline tests/try_scraper/test_invalid_changeset.py]: nit: please fix the comment just for clarity.

File: tests/try_scraper/test_try_scraper.py [hunk 25]
```diff
@@ -0,0 +1,74 @@
...
+              'changeset': '8fcac92cfcad'},
+     'target': '8fcac92cfcad-firefox-38.0a1.en-US.win32.installer.exe',
+     'target_url': 'firefox/try-builds/test-user@mozilla.com-8fcac92cfcad/'
+                   'try-win32/firefox-38.0a1.en-US.win32.installer.exe'},
+    # -a firefox -p win64 --changeset=8fcac92cfcad
+    {'args': {'platform': 'win64',
+              'changeset': '8fcac92cfcad'},
+     'target': '8fcac92cfcad-firefox-38.0a1.en-US.win64-x86_64.installer.exe',
+     'target_url': 'firefox/try-builds/test-user@mozilla.com-8fcac92cfcad/'
+                   'try-win64/firefox-38.0a1.en-US.win64-x86_64.installer.exe'},
```
Comment (whimboo) [inline tests/try_scraper/test_try_scraper.py]: This is using the old format of the filename. Given that try builds don't survive that long, change that to `win64.installer`. I don't see a reason to test for `x86_64`.

**Comment by whimboo** [comment]
This looks great! Just some minor changes necessary and we can get it landed.

**Comment by Nebelhom** [comment]
@whimboo 

I told @EricRahm that all commits will be squashed once the PR has been accepted before merging. I must have miscommunicated it, which is why he did after each change.

Apologies to both of you for the confusion.

File: tests/try_scraper/test_invalid_changeset.py (hunk 24 above)
Comment (EricRahm) [inline tests/try_scraper/test_invalid_changeset.py]: Will update.

File: tests/try_scraper/test_try_scraper.py (hunk 25 above)
Comment (EricRahm) [inline tests/try_scraper/test_try_scraper.py]: Will update and rename.

**Comment by EricRahm** [comment]
@Nebelhom @whimboo re: squash, each time I assumed I was done :( I'll wait for an official request this time!

**Comment by whimboo** [comment]
This looks fine to me now. I'm just waiting for Travis to report a pass. I will squash and merge it shortly after.

**Comment by whimboo** [comment]
This PR has been squashed and merged to master as https://github.com/mozilla/mozdownload/commit/f42b772f7b272559a22c73e18bd06510deac4576. Thanks @EricRahm for working on it!
